cpp/src/io/json/json_tree.cu

```diff
@@ -380,6 +389,101 @@ tree_meta_t get_tree_representation(device_span<PdaTokenT const> tokens,
     stream);
   CUDF_EXPECTS(node_range_out_end - node_range_out_it == num_nodes, "node range count mismatch");
 
+  // Extract Struct, List range_end:
+  // 1. Extract Struct, List - begin & end separately, their token ids
+  // 2. push, pop to get levels
+  // 3. copy first child's parent token_id, also translate to node_id
+  // 4. propagate to siblings using levels, parent token id. (segmented scan)
+  // 5. scatter to node_range_end for only nested end tokens.
+  if (is_strict_nested_boundaries) {
+    // Whether the token is nested
+    auto const is_nested = [] __device__(PdaTokenT const token) -> bool {
+      switch (token) {
+        case token_t::StructBegin:
+        case token_t::StructEnd:
+        case token_t::ListBegin:
+        case token_t::ListEnd: return true;
+        default: return false;
+      };
+    };
+    auto const num_nested =
+      thrust::count_if(rmm::exec_policy(stream), tokens.begin(), tokens.end(), is_nested);
+    rmm::device_uvector<TreeDepthT> token_levels(num_nested, stream);
+    rmm::device_uvector<NodeIndexT> token_id(num_nested, stream);
+    rmm::device_uvector<NodeIndexT> parent_node_ids(num_nested, stream);
+    auto const push_pop_it = thrust::make_transform_iterator(
+      tokens.begin(),
+      cuda::proclaim_return_type<cudf::size_type>([] __device__(PdaTokenT const token) {
+        size_type const is_begin = token == token_t::StructBegin or token == token_t::ListBegin;
+        size_type const is_end   = token == token_t::StructEnd or token == token_t::ListEnd;
```

Comment:
Is this implicitly casting `bool` to `size_type`? Should we make that explicit? See also my other comment about rewriting this lambda.